api/v1/nonclusterhost_types.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2024 Tigera, Inc. All rights reserved.
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NonClusterHostSpec enables non-cluster hosts to connect to a cluster.
+type NonClusterHostSpec struct {
+	// Location of the log ingestion point for non-cluster hosts. example: https://1.2.3.4:443
+	Endpoint string `json:"endpoint"`
```

Comment:
This seems like a required field. Should we add:
``json:"endpoint" validate:"required"``

Reply:
It might be worth writing a custom validator, so that the string input that does not match the a predefined regex pattern prints an error at the time of creation.

Reply:
Validation pattern is added to the endpoint field.

Reply:
I wonder if the field name should be made more specific. Like logstorageEndpoint. I imagine we may have a need for a second endpoint for other purposes, for example an endpoint that will be used for connecting to typha.

Reply:
Since NonClusterHostSpec is not a required field, please mind that it is still possible to create this resource without the Endpoint field. Does that have any implications for the feature? Should the controller still validate that the field exists?

Reply:
NonClusterHostSpec isn't required, but it's also not a pointer, which means it will be present regardless and so this field should always be defaulted by the kubebuilder validation IIUC

Reply:
I applied the CRD to my kind cluster:

```
$ k apply -f nonclust.yaml 
The NonClusterHost "default" is invalid: spec.endpoint: Required value

 1   12:17:55   ~/dev/kind    
$ cat nonclust.yaml 
kind: NonClusterHost
apiVersion: operator.tigera.io/v1
metadata:
  name: default
spec: {}
```

But then:

```
$ k apply -f nonclust.yaml 
nonclusterhost.operator.tigera.io/default configured
 0   12:18:26   ~/dev/kind    
$ cat nonclust.yaml 
kind: NonClusterHost
apiVersion: operator.tigera.io/v1
metadata:
  name: default
spec: null
```

Reply:
hmm, guess I was wrong about that. Surprised that `spec: null` is valid at all considering the spec isn't a pointer.

Reply:
It is identical to this: 
```
kind: NonClusterHost
apiVersion: operator.tigera.io/v1
metadata:
  name: default
```

Reply:
`spec.endpoint` is validated in the manager controller. When NonClusterHost resource is applied to the cluster and endpoint filed is an invalid URL, manager controller will be degraded.